### ISSUE TITLE
Use Dedicated Macro to Signal 1-based versus 0-based Array Indexing

### ISSUE THREAD

**Issue by john-boyer-phd** [graph-algorithms/edge-addition-planarity-suite#24]
The original implementation stored vertices and edges using 0-based array indexing. This resulted in a definition of -1 for the `NIL` index (indicating no vertex or index). A significant performance increase was achieved by changing `NIL` to 0 because then every single test for `NIL` became an equality comparison with 0.

Hence, the value of the `NIL` macro is used in `#ifdef` tests throughout the project to decide whether to compile 1-based or 0-based code. 

The goal in this issue is to define a macro, such as `#define ARRAY_BASE_INDEX 1`, or `0`, and then let that setting drive the decisions about what code to compile, including what value is chosen for `NIL`.

**Comment by wbkboyer** [comment]
Another option would be to use a compiler flag (i.e. `-D ONEBASED` ([`gcc` - Preprocessor Options](https://gcc.gnu.org/onlinedocs/gcc/Preprocessor-Options.html#index-D-1) and [`clang` - `/D` Macro](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-D-macro)) or `/D ONEBASED` ([MSVC `cl` - `/D` (Preprocessor Definitions)](https://learn.microsoft.com/en-us/cpp/build/reference/d-preprocessor-definitions?view=msvc-170))), which according to the documentation is equivalent to adding the preprocessor directive `#define ONEBASED 1` to the beginning of a source code file. This would mean that we could use `#ifdef ONEBASED` statements, for example conditioning the value of `NIL` in `appconst.h` or how we choose to initialize memory for the graph algorithm extensions.

A benefit of this approach, as opposed to adding `#define ARRAY_BASE_INDEX 1`/`0` to `appconst.h` is that we no longer run the risk of contributors erroneously including their local changes to  switch between 0- and 1-based array indexing (the default should be 1-based due to the aforementioned speed boosts).

The only thorn I could think is that since 1-based array indexing should be considered the *default*, it feels bad to require one to include the preprocessor option compiler flag in the build command to enforce this default (i.e. both in `devEnvSetupAndDefaults/.vscode/tasks.json` and adding the [`planarity_CPPFLAGS`](https://www.gnu.org/software/automake/manual/html_node/Flag-Variables-Ordering.html)` = -D ONEBASED` user variable to `Makefile.am`). The alternative would be to use `#ifdef ZEROBASED` statements (flips the order of some of the current preprocessor conditionals), where by default the build command wouldn't include any preprocessor option, and users would have to modify their local, untracked `.vscode/tasks.json` to compile with `-D ZEROBASED`.

**Comment by john-boyer-phd** [comment]
I agree with, like, and prefer the ZEROBASED option because ONEBASED should be the default unless somebody has a good reason to go back and test zero-based behavior for some ancient legacy reason. Also means there should be a slight change to the wiki that describes how to trigger this change from default 1-based to ZEROBASED.
